docs/content/2.config.md
---
icon: i-ph-gear-six-duotone
description: 'Customize your Nitro app with a configuration file.'
aside: false
---

In order to customize Nitro's behavior, we create a file named `nitro.config.ts`.

::code-group
```ts [nitro.config.ts]
export default defineNitroConfig({
  // options
});
```
```ts [nuxt.config.ts]
export default defineNuxtConfig({
  nitro: {
    // options
  }
})
```
::

## General

### `preset`

Use `preset` option `NITRO_PRESET` environment variable for custom **production** preset.

Preset for development mode is always `nitro-dev` and default `node-server` for production building a standalone Node.js server.

The preset will automatically be detected when the `preset` option is not set and running in known environments.

### `logLevel`

- Default: `3`{lang=ts} (`1`{lang=ts} when the testing environment is detected)

Log verbosity level. See [unjs/consola#level](https://github.com/unjs/consola/#level) for more information.

### `runtimeConfig`

- Default: `{ nitro: { ... }, ...yourOptions }`{lang=ts}

Server runtime configuration.

**Note:**: `nitro` namespace is reserved.

## Features

### `experimental`

- Default: `{}`

Enable experimental features.

#### `openAPI`

Enable `/_nitro/swagger` and `/_nitro/openapi.json` endpoints.

#### `wasm`

Enable WASM support

#### `legacyExternals`

When enabled, legacy (unstable) experimental rollup externals algorithm will be used.

### `future`

- Default: `{}`

New features pending for a major version to avoid breaking changes.

#### `nativeSWR`

Uses built-in SWR functionality (using caching layer and storage) for Netlify and Vercel presets instead of falling back to ISR behavior.

### `storage`

- Default: `{}`

Storage configuration, read more in the [Storage Layer](/guide/storage) section.

### `timing`

- Default: `false`{lang=ts}

Enable timing information:

- Nitro startup time log
- `Server-Timing` header on HTTP responses

### `renderer`

Path to main render (file should export an event handler as default)

### `serveStatic`

- Type: `boolean` | `'node'`{lang=ts} | `'deno'`{lang=ts}
- Default: depends of the deployment preset used.

Serve `public/` assets in production.

**Note:** It is highly recommended that your edge CDN (nginx, apache, cloud) serves the `.output/public/` directory instead to enable compression and higher lever caching.

### `noPublicDir`

- Default: `false`{lang=ts}

If enabled, disabled `.output/public` directory creation. Skipping to copy `public/` dir and also disables pre-renderering.

### `publicAssets`

Public asset directories to serve in development and bundle in production.

If a `public/` directory is detected, it will be added by default, but you can add more by yourself too!

It's possible to set Cache-Control headers for assets using the `maxAge` option:
```ts
  publicAssets: [
    {
      baseURL: "images",
      dir: "public/images",
      maxAge: 60 * 60 * 24 * 7, // 7 days
    },
  ],
```

The config above generates the following header in the assets under `public/images/` folder:

`cache-control: public, max-age=604800, immutable`

### `compressPublicAssets`

- Default: `{ gzip: false, brotli: false }`{lang=ts}

If enabled, Nitro will generate a pre-compressed (gzip and/or brotli) version of supported types of public assets and prerendered routes
larger than 1024 bytes into the public directory. The best compression level is used. Using this option you can support zero overhead asset compression without using a CDN.

::callout
#summary
List of compressible MIME types:
#content
- `application/dash+xml`
- `application/eot`
- `application/font`
- `application/font-sfnt`
- `application/javascript`
- `application/json`
- `application/opentype`
- `application/otf`
- `application/pkcs7-mime`
- `application/protobuf`
- `application/rss+xml`
- `application/truetype`
- `application/ttf`
- `application/vnd.apple.mpegurl`
- `application/vnd.mapbox-vector-tile`
- `application/vnd.ms-fontobject`
- `application/xhtml+xml`
- `application/xml`
- `application/x-font-opentype`
- `application/x-font-truetype`
- `application/x-font-ttf`
- `application/x-httpd-cgi`
- `application/x-javascript`
- `application/x-mpegurl`
- `application/x-opentype`
- `application/x-otf`
- `application/x-perl`
- `application/x-ttf`
- `font/eot`
- `font/opentype`
- `font/otf`
- `font/ttf`
- `image/svg+xml`
- `text/css`
- `text/csv`
- `text/html`
- `text/javascript`
- `text/js`
- `text/plain`
- `text/richtext`
- `text/tab-separated-values`
- `text/xml`
- `text/x-component`
- `text/x-java-source`
- `text/x-script`
- `vnd.apple.mpegurl`
::

### `serverAssets`

Assets can be accessed in server logic and bundled in production. [Read more](/guide/assets#server-assets).

### `devServer`

- Default: `{ watch: [] }`{lang=ts}

Dev server options. You can use `watch` to make the dev server reload if any file changes in specified paths.

### `watchOptions`

Watch options for development mode. See [chokidar](https://github.com/paulmillr/chokidar) for more information.

### `imports`

Auto import options. See [unjs/unimport](https://github.com/unjs/unimport) for more information.

### `plugins`

- Default: `[]`

An array of paths to nitro plugins. They will be executed by order on the first initialization.

Note that Nitro auto-register the plugins in the `plugins/` directory, [learn more](/guide/plugins).

### `virtual`

- Default: `{}`

A map from dynamic virtual import names to their contents or an (async) function that returns it.

## Routing

### `baseURL`

Default: `/`{lang=ts} (or `NITRO_APP_BASE_URL` environment variable if provided)

Server's main base URL.

### `handlers`

Server handlers and routes.

If `routes/`, `api/` or `middleware/` directories exist, they will be automatically added to the handlers array.

### `devHandlers`

Regular handlers refer to the path of handlers to be imported and transformed by rollup.

There are situations in that we directly want to provide a handler instance with programmatic usage.

We can use `devHandlers` but note that they are **only available in development mode** and **not in production build**.

For example:

```ts
import { defineEventHandler } from 'h3'

export default defineNitroConfig({
  devHandlers: [
    {
      route: '/',
      handler: defineEventHandler((event) => {
       console.log(event)
      })
    }
  ]
})
```

::callout{ icon="i-ph-info-duotone" color="blue" }
Note that `defineEventHandler` is a helper function from [`h3`](https://github.com/unjs/h3) library.
::

### `devProxy`

Proxy configuration for development server.

You can use this option to override development server routes and proxy-pass requests.

```js
{
  devProxy: {
    '/proxy/test': 'http://localhost:3001',
    '/proxy/example': { target: 'https://example.com', changeOrigin: true }
  }
}
```

See [unjs/httpxy](https://github.com/unjs/httpxy) for all available target options.

### `errorHandler`

Path to a custom runtime error handler. Replacing nitro's built-in error page.
The error handler is given an `H3Error` and `H3Event`. If the handler returns a promise it is awaited.
The handler is expected to send a response of its own.
Below is an example where a plain-text response is returned using h3's functions.

**Example:**

```js [nitro.config]
export default defineNitroConfig({
  errorHandler: "~/error",
});
```

```js [error.ts]
export default defineNitroErrorHandler((error, event) => {
  setResponseHeader('Content-Type', 'text/plain')
  return send(event, '[custom error handler] ' + error.stack)
});
```

### `routeRules`

**🧪 Experimental!**

Route options. It is a map from route pattern (following [unjs/radix3](https://github.com/unjs/radix3#route-matcher)) to route options.

When `cache` option is set, handlers matching pattern will be automatically wrapped with `defineCachedEventHandler`.

See the [Cache API](/guide/cache) for all available cache options.

::callout
`swr: true|number` is shortcut for `cache: { swr: true, maxAge: number }`
::

**Example:**

```js
routeRules: {
  '/blog/**': { swr: true },
  '/blog/**': { swr: 600 },
  '/blog/**': { static: true },
  '/blog/**': { cache: { /* cache options*/ } },
  '/assets/**': { headers: { 'cache-control': 's-maxage=0' } },
  '/api/v1/**': { cors: true, headers: { 'access-control-allow-methods': 'GET' } },
  '/old-page': { redirect: '/new-page' }, // uses status code 307 (Temporary Redirect)
  '/old-page2': { redirect: { to:'/new-page2', statusCode: 301 } },
  '/proxy/example': { proxy: 'https://example.com' },
  '/proxy/**': { proxy: '/api/**' },
}
```

### `prerender`

Default:

```ts
{
  autoSubfolderIndex: true,
  concurrency: 1,
  interval: 0,
  failOnError: false,
  crawlLinks: false,
  ignore: [],
  routes: [],
  retries: 3,
  retryDelay: 500
}
```

Prerendered options. Any route specified will be fetched during the build and copied to the `.output/public` directory as a static asset.

Any route that starts with a prefix listed in `ignore` will be ignored.

If `crawlLinks` option is set to `true`, nitro starts with `/` by default (or all routes in `routes` array) and for HTML pages extracts `<a>` tags and prerender them as well.

You can set `failOnError` option to `true` to stop the CI when an error if Nitro could not prerender a route.

The `interval` and `concurrency` options lets you control the speed of pre-rendering, can be useful to avoid hitting some rate-limit if you call external APIs.

Set `autoSubfolderIndex` lets you control how to generate the files in the `.output/public` directory:

```bash
# autoSubFolderIndex: true (default)
/about -> .output/public/about/index.html
# autoSubFolderIndex: false
/about -> .output/public/about.html
```

This option is useful when your hosting provider does not give you an option regarding the trailing slash.

The prerenderer will attempt to render pages 3 times with a delay of 500ms. Use `retry` and `retryDelay` to change this behavior.

## Directories

### `rootDir`

Project main directory

### `srcDir`

Project source directory. Same as `rootDir` unless specified. Helpful to move code into `src/`.

### `scanDirs`

- Default: (source directory when empty array)

List of directories to scan and auto-register files, such as API routes.

### `buildDir`

- Default: `.nitro`

nitro's temporary working directory for generating build-related files.

### `output`

- Default: `{ dir: '.output', serverDir: '.output/server', publicDir: '.output/public' }`

Output directories for production bundle.

## Advanced

### `dev`

- Default: `true` for development and `false` for production.

**⚠️ Caution! This is an advanced configuration. things can go wrong if misconfigured.**

### `typescript`

Default: `{ generateTsConfig: true }`

### `nodeModulesDirs`

**⚠️ Caution! This is an advanced configuration. things can go wrong if misconfigured.**

Additional `node_modules` to search when resolving a module. By default user directory is added.

### `hooks`

**⚠️ Caution! This is an advanced configuration. things can go wrong if misconfigured.**

nitro hooks. See [unjs/hookable](https://github.com/unjs/hookable) for more information.

### `commands`

**⚠️ Caution! This is an advanced configuration. things can go wrong if misconfigured.**

Preview and deploy command hints are usually filled by deployment presets.

### `devErrorHandler`

**⚠️ Caution! This is an advanced configuration. things can go wrong if misconfigured.**

A custom error handler function for development errors.

## Rollup

### `rollupConfig`

Additional rollup configuration.

### `entry`

Rollup entry.

### `unenv`

Options for [unjs/unenv](https://github.com/unjs/unenv/) preset.

### `alias`

Rollup aliases options.

### `minify`

- Default: `false`

Minify bundle.

### `inlineDynamicImports`

Avoid creating chunks.

### `sourceMap`

Enable source-map generation. See [options](https://rollupjs.org/configuration-options/#output-sourcemap)
- Default: `true`

### `node`

Specify whether the build is used for Node.js or not. If set to `false`, nitro tried to mock Node.js dependencies using [unjs/unenv](https://github.com/unjs/unenv) and adjust its behavior.

### `analyze`

If enabled, will analyze server bundle after build using [rollup-plugin-visualizer](https://github.com/btd/rollup-plugin-visualizer). You can also pass your custom options.

### `moduleSideEffects`

Default: `['unenv/runtime/polyfill/', 'node-fetch-native/polyfill']`

Rollup specific option. Specifies module imports that have side-effects

### `replace`

Rollup specific option.

### `commonJS`

Rollup specific option. Specifies additional configuration for the rollup CommonJS plugin.

## Preset options

### `firebase`

The options for the firebase functions preset. See [Preset Docs](/deploy/providers/firebase#options)

### `vercel`

The options for the vercel preset. See [Preset Docs](/deploy/providers/vercel)

### `cloudflare`

The options for the cloudflare preset. See [Preset Docs](/deploy/providers/cloudflare)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## v2.8.0

[compare changes](https://github.com/unjs/nitro/compare/v2.7.1...v2.8.0)

### 🚀 Enhancements

- Support modules to extend nitro builder ([#1789](https://github.com/unjs/nitro/pull/1789))
- Support `inline` mode for `serveStatic` ([#1864](https://github.com/unjs/nitro/pull/1864))
- Add experimental `winterjs` preset ([#1863](https://github.com/unjs/nitro/pull/1863))
- Allow setting preset using `SERVER_PRESET` environment variable ([#1870](https://github.com/unjs/nitro/pull/1870))
- **cloudflare-pages, cloudflare-module:** Enable code splitting by default ([#1905](https://github.com/unjs/nitro/pull/1905))
- Support `framework` config ([#1843](https://github.com/unjs/nitro/pull/1843))
- Add aws amplify hosting preset ([#1912](https://github.com/unjs/nitro/pull/1912))
- Upgrade to rollup v4 ([#1927](https://github.com/unjs/nitro/pull/1927))
- Improve `nitro.json` build info ([#1930](https://github.com/unjs/nitro/pull/1930))
- **aws-amplify:** Support static builds ([#1933](https://github.com/unjs/nitro/pull/1933))
- Add `defineNitroErrorHandler` type helper ([#1923](https://github.com/unjs/nitro/pull/1923))

### 🩹 Fixes

- **cloudflare-pages:** Filter out overlapping public assets dirs ([#1859](https://github.com/unjs/nitro/pull/1859))
- **winterjs:** 0.1.7 compatibility ([#1876](https://github.com/unjs/nitro/pull/1876))
- **scanner:** Allow having http method as part of the route name ([#1895](https://github.com/unjs/nitro/pull/1895))
- Generate route types with resolve path and stripped extension ([#1897](https://github.com/unjs/nitro/pull/1897))
- Scan dirs for unimport on initialization ([#1908](https://github.com/unjs/nitro/pull/1908))
- **aws-amplify:** Register and auto detect ([0f38eb6f](https://github.com/unjs/nitro/commit/0f38eb6f))
- **prerender:** Decode generated routes ([#1914](https://github.com/unjs/nitro/pull/1914))
- Disabled public asset handler name is `null` ([b7a6a1a6](https://github.com/unjs/nitro/commit/b7a6a1a6))
- **prerender:** Decode uris in headers ([#1932](https://github.com/unjs/nitro/pull/1932))
- **error:** Respect `accept: text/html` request header ([#1921](https://github.com/unjs/nitro/pull/1921))
- **prerender:** Filter encoded links starting with `#` ([#1936](https://github.com/unjs/nitro/pull/1936))

### 💅 Refactors

- **deno-deploy:** Use `Deno.serve` ([#1879](https://github.com/unjs/nitro/pull/1879))
- Split preset types ([#1910](https://github.com/unjs/nitro/pull/1910))

### 📖 Documentation

- **deploy:** Split runtimes from providers ([#1865](https://github.com/unjs/nitro/pull/1865))
- **routing:** Add note about middleware returns ([#1884](https://github.com/unjs/nitro/pull/1884))
- Update edge links to nightly ([#1902](https://github.com/unjs/nitro/pull/1902))
- **deploy:** Add deno to runtime list ([#1899](https://github.com/unjs/nitro/pull/1899))
- **cloudflare:** Updatge wrangler link ([#1913](https://github.com/unjs/nitro/pull/1913))
- **aws-amplify:** Add custom `amplify.yml` ([f412af16](https://github.com/unjs/nitro/commit/f412af16))
- **amplify:** Remove cache from config ([7bbb723e](https://github.com/unjs/nitro/commit/7bbb723e))
- Hide winterjs due to unstability ([048c8ff0](https://github.com/unjs/nitro/commit/048c8ff0))
- **aws-amplify:** Update regions ([#1931](https://github.com/unjs/nitro/pull/1931))
- **aws-amplify:** Add step to enable SSR logging and remove experimental ([#1934](https://github.com/unjs/nitro/pull/1934))
- **aws-amplify:** Remove workaround banner ([#1935](https://github.com/unjs/nitro/pull/1935))

### 🌊 Types

- Export `SerializeTuple` and `SerializeObject` ([#1907](https://github.com/unjs/nitro/pull/1907))

### 🏡 Chore

- **release:** V2.7.1 ([2d803a3c](https://github.com/unjs/nitro/commit/2d803a3c))
- Add jsdoc for `cache.maxAge` ([#1878](https://github.com/unjs/nitro/pull/1878))
- Update lockfile ([dbff232e](https://github.com/unjs/nitro/commit/dbff232e))
- **docs:** Update dependencies ([28867009](https://github.com/unjs/nitro/commit/28867009))
- **docs:** Fix prerender issues ([d2a7d84d](https://github.com/unjs/nitro/commit/d2a7d84d))
- Update unstorage ([110bc4a3](https://github.com/unjs/nitro/commit/110bc4a3))
- Update unstorage ([441db6ca](https://github.com/unjs/nitro/commit/441db6ca))
- Fix typos in aws amplify ([963716dd](https://github.com/unjs/nitro/commit/963716dd))
- Update lockfile ([926bcef4](https://github.com/unjs/nitro/commit/926bcef4))

### ✅ Tests

- Hide winterjs due to unstability ([1b50a9a1](https://github.com/unjs/nitro/commit/1b50a9a1))

### 🤖 CI

- Disable codecov and codeql ([fd4671ce](https://github.com/unjs/nitro/commit/fd4671ce))

### ❤️ Contributors

- Pooya Parsa ([@pi0](http://github.com/pi0))
- Siegerts <stephen.siegert@gmail.com>
- Passionate-bram 
- Daniel Roe <daniel@roe.dev>
- Neil Richter ([@noook](http://github.com/noook))
- Xin Du (Clark) <clark.duxin@gmail.com>
- Anthony Fu <anthonyfu117@hotmail.com>
- Dario Piotrowicz ([@dario-piotrowicz](http://github.com/dario-piotrowicz))
- Kalwabed Rizki 
- Thunfisch987 
- Oof2win2 <honza.koco44@gmail.com>
- Markthree ([@markthree](http://github.com/markthree))
- Estéban ([@Barbapapazes](http://github.com/Barbapapazes))

## v2.7.1

[compare changes](https://github.com/unjs/nitro/compare/v2.7.0...v2.7.1)

### 🩹 Fixes

- Recursively simplify returned objects ([#1847](https://github.com/unjs/nitro/pull/1847))
- **cache:** Fix `etag` and `last-modified` values ([#1855](https://github.com/unjs/nitro/pull/1855))
- **cache:** Invalidate wrongly cached  handler entities ([#1857](https://github.com/unjs/nitro/pull/1857))

### 🏡 Chore

- Update dependencies ([7a6c61db](https://github.com/unjs/nitro/commit/7a6c61db))
- Update lockfile ([9b5ed9a1](https://github.com/unjs/nitro/commit/9b5ed9a1))

src/options.ts
import { resolve, join } from "pathe";
import { loadConfig, watchConfig, WatchConfigOptions } from "c12";
import { klona } from "klona/full";
import { camelCase } from "scule";
import { defu } from "defu";
import { resolveModuleExportNames } from "mlly";
import escapeRE from "escape-string-regexp";
import { withLeadingSlash, withoutTrailingSlash, withTrailingSlash } from "ufo";
import { isTest, isDebug, nodeMajorVersion, provider } from "std-env";
import { findWorkspaceDir } from "pkg-types";
import consola from "consola";
import { version } from "../package.json";
import {
  resolvePath,
  resolveFile,
  detectTarget,
  provideFallbackValues,
} from "./utils";
import type {
  NitroConfig,
  NitroOptions,
  NitroRouteConfig,
  NitroRouteRules,
  NitroRuntimeConfig,
} from "./types";
import { runtimeDir, pkgDir } from "./dirs";
import * as _PRESETS from "./presets";
import { nitroImports } from "./imports";

const NitroDefaults: NitroConfig = {
  // General
  debug: isDebug,
  timing: isDebug,
  logLevel: isTest ? 1 : 3,
  runtimeConfig: { app: {}, nitro: {} },
  appConfig: {},
  appConfigFiles: [],

  // Dirs
  scanDirs: [],
  buildDir: ".nitro",
  output: {
    dir: "{{ rootDir }}/.output",
    serverDir: "{{ output.dir }}/server",
    publicDir: "{{ output.dir }}/public",
  },

  // Features
  experimental: {},
  future: {},
  storage: {},
  devStorage: {},
  bundledStorage: [],
  publicAssets: [],
  serverAssets: [],
  plugins: [],
  tasks: {},
  imports: {
    exclude: [],
    dirs: [],
    presets: nitroImports,
    virtualImports: ["#imports"],
  },
  virtual: {},
  compressPublicAssets: false,
  ignore: [],

  // Dev
  dev: false,
  devServer: { watch: [] },
  watchOptions: { ignoreInitial: true },
  devProxy: {},

  // Logging
  logging: {
    compressedSizes: true,
    buildSuccess: true,
  },

  // Routing
  baseURL: process.env.NITRO_APP_BASE_URL || "/",
  handlers: [],
  devHandlers: [],
  errorHandler: "#internal/nitro/error",
  routeRules: {},
  prerender: {
    autoSubfolderIndex: true,
    concurrency: 1,
    interval: 0,
    retry: 3,
    retryDelay: 500,
    failOnError: false,
    crawlLinks: false,
    ignore: [],
    routes: [],
  },

  // Rollup
  alias: {
    "#internal/nitro": runtimeDir,
  },
  unenv: {},
  analyze: false,
  moduleSideEffects: [
    "unenv/runtime/polyfill/",
    "node-fetch-native/polyfill",
    "node-fetch-native/dist/polyfill",
    resolve(runtimeDir, "polyfill/"),
  ],
  replace: {},
  node: true,
  sourceMap: true,
  esbuild: {
    options: {
      jsxFactory: "h",
      jsxFragment: "Fragment",
    },
  },

  // Advanced
  typescript: {
    strict: false,
    generateTsConfig: true,
    tsconfigPath: "types/tsconfig.json",
    internalPaths: false,
    tsConfig: {},
  },
  nodeModulesDirs: [],
  hooks: {},
  commands: {},

  // Framework
  framework: {
    name: "nitro",
    version,
  },
};

export interface LoadConfigOptions {
  watch?: boolean;
  c12?: WatchConfigOptions;
}

export async function loadOptions(
  configOverrides: NitroConfig = {},
  opts: LoadConfigOptions = {}
): Promise<NitroOptions> {
  // Preset
  let presetOverride =
    (configOverrides.preset as string) ||
    process.env.NITRO_PRESET ||
    process.env.SERVER_PRESET;
  if (configOverrides.dev) {
    presetOverride = "nitro-dev";
  }

  // Load configuration and preset
  configOverrides = klona(configOverrides);
  // @ts-ignore
  globalThis.defineNitroConfig = globalThis.defineNitroConfig || ((c) => c);
  const c12Config = await (opts.watch ? watchConfig : loadConfig)(<
    WatchConfigOptions
  >{
    name: "nitro",
    cwd: configOverrides.rootDir,
    dotenv: configOverrides.dev,
    extend: { extendKey: ["extends", "preset"] },
    overrides: {
      ...configOverrides,
      preset: presetOverride,
    },
    defaultConfig: {
      preset: detectTarget({ static: configOverrides.static }),
    },
    defaults: NitroDefaults,
    jitiOptions: {
      alias: {
        nitropack: "nitropack/config",
        "nitropack/config": "nitropack/config",
      },
    },
    resolve(id: string) {
      const presets = _PRESETS as any as Map<string, NitroConfig>;
      let matchedPreset = presets[camelCase(id)] || presets[id];
      if (!matchedPreset) {
        return null;
      }
      if (typeof matchedPreset === "function") {
        matchedPreset = matchedPreset();
      }
      return {
        config: matchedPreset,
      };
    },
    ...opts.c12,
  });
  const options = klona(c12Config.config) as NitroOptions;
  options._config = configOverrides;
  options._c12 = c12Config;

  options.preset =
    presetOverride ||
    (c12Config.layers.find((l) => l.config.preset)?.config.preset as string) ||
    detectTarget({ static: options.static });

  options.rootDir = resolve(options.rootDir || ".");
  options.workspaceDir = await findWorkspaceDir(options.rootDir).catch(
    () => options.rootDir
  );
  options.srcDir = resolve(options.srcDir || options.rootDir);
  for (const key of ["srcDir", "publicDir", "buildDir"]) {
    options[key] = resolve(options.rootDir, options[key]);
  }

  // Add aliases
  options.alias = {
    ...options.alias,
    "~/": join(options.srcDir, "/"),
    "@/": join(options.srcDir, "/"),
    "~~/": join(options.rootDir, "/"),
    "@@/": join(options.rootDir, "/"),
  };

  // Resolve possibly template paths
  if (!options.static && !options.entry) {
    throw new Error(
      `Nitro entry is missing! Is "${options.preset}" preset correct?`
    );
  }
  if (options.entry) {
    options.entry = resolvePath(options.entry, options);
  }
  options.output.dir = resolvePath(
    options.output.dir || NitroDefaults.output.dir,
    options,
    options.rootDir
  );
  options.output.publicDir = resolvePath(
    options.output.publicDir || NitroDefaults.output.publicDir,
    options,
    options.rootDir
  );
  options.output.serverDir = resolvePath(
    options.output.serverDir || NitroDefaults.output.serverDir,
    options,
    options.rootDir
  );

  options.nodeModulesDirs.push(resolve(options.workspaceDir, "node_modules"));
  options.nodeModulesDirs.push(resolve(options.rootDir, "node_modules"));
  options.nodeModulesDirs.push(resolve(pkgDir, "node_modules"));
  options.nodeModulesDirs.push(resolve(pkgDir, "..")); // pnpm
  options.nodeModulesDirs = [
    ...new Set(
      options.nodeModulesDirs.map((dir) => resolve(options.rootDir, dir))
    ),
  ];

  // Resolve scanDirs
  options.scanDirs.unshift(options.srcDir);
  options.scanDirs = options.scanDirs.map((dir) =>
    resolve(options.srcDir, dir)
  );
  options.scanDirs = [...new Set(options.scanDirs)];

  if (
    options.imports &&
    Array.isArray(options.imports.exclude) &&
    options.imports.exclude.length === 0
  ) {
    // Exclude .git and buildDir by default
    options.imports.exclude.push(/[/\\]\.git[/\\]/);
    options.imports.exclude.push(options.buildDir);

    // Exclude all node modules that are not a scanDir
    const scanDirsInNodeModules = options.scanDirs
      .map((dir) => dir.match(/(?<=\/)node_modules\/(.+)$/)?.[1])
      .filter(Boolean);
    options.imports.exclude.push(
      scanDirsInNodeModules.length > 0
        ? new RegExp(
            `node_modules\\/(?!${scanDirsInNodeModules
              .map((dir) => escapeRE(dir))
              .join("|")})`
          )
        : /[/\\]node_modules[/\\]/
    );
  }

  // Add h3 auto imports preset
  if (options.imports) {
    const h3Exports = await resolveModuleExportNames("h3", {
      url: import.meta.url,
    });
    options.imports.presets.push({
      from: "h3",
      imports: h3Exports.filter((n) => !/^[A-Z]/.test(n) && n !== "use"),
    });
  }

  // Auto imports from utils dirs
  if (options.imports) {
    options.imports.dirs.push(
      ...options.scanDirs.map((dir) => join(dir, "utils/*"))
    );
  }

  // Normalize app.config file paths
  options.appConfigFiles = options.appConfigFiles
    .map((file) => resolveFile(resolvePath(file, options)))
    .filter(Boolean);

  // Detect app.config from scanDirs
  for (const dir of options.scanDirs) {
    const configFile = resolveFile("app.config", dir);
    if (configFile && !options.appConfigFiles.includes(configFile)) {
      options.appConfigFiles.push(configFile);
    }
  }

  // Backward compatibility for options.routes
  options.routeRules = defu(options.routeRules, (options as any).routes || {});

  // Normalize route rules
  options.routeRules = normalizeRouteRules(options);

  options.baseURL = withLeadingSlash(withTrailingSlash(options.baseURL));

  // Normalize runtime config
  options.runtimeConfig = normalizeRuntimeConfig(options);

  for (const publicAsset of options.publicAssets) {
    publicAsset.dir = resolve(options.srcDir, publicAsset.dir);
    publicAsset.baseURL = withLeadingSlash(
      withoutTrailingSlash(publicAsset.baseURL || "/")
    );
  }

  for (const serverAsset of options.serverAssets) {
    serverAsset.dir = resolve(options.srcDir, serverAsset.dir);
  }

  // Build-only storage
  const fsMounts = {
    root: resolve(options.rootDir),
    src: resolve(options.srcDir),
    build: resolve(options.buildDir),
    cache: resolve(options.buildDir, "cache"),
  };
  for (const p in fsMounts) {
    options.devStorage[p] = options.devStorage[p] || {
      driver: "fs",
      readOnly: p === "root" || p === "src",
      base: fsMounts[p],
    };
  }

  // Runtime storage
  if (
    options.dev &&
    options.storage.data === undefined &&
    options.devStorage.data === undefined
  ) {
    options.devStorage.data = {
      driver: "fs",
      base: resolve(options.rootDir, ".data/kv"),
    };
  } else if (options.node && options.storage.data === undefined) {
    options.storage.data = {
      driver: "fsLite",
      base: "./.data/kv",
    };
  }

  // Resolve plugin paths
  options.plugins = options.plugins.map((p) => resolvePath(p, options));

  // Export conditions
  options.exportConditions = _resolveExportConditions(
    options.exportConditions,
    { dev: options.dev, node: options.node }
  );

  // Add open-api endpoint
  if (options.dev && options.experimental.openAPI) {
    options.handlers.push({
      route: "/_nitro/openapi.json",
      handler: "#internal/nitro/routes/openapi",
    });
    options.handlers.push({
      route: "/_nitro/swagger",
      handler: "#internal/nitro/routes/swagger",
    });
  }

  // Native fetch
  if (options.experimental.nodeFetchCompat === undefined) {
    options.experimental.nodeFetchCompat = nodeMajorVersion < 18;
    if (options.experimental.nodeFetchCompat && provider !== "stackblitz") {
      consola.warn(
        "Node fetch compatibility is enabled. Please consider upgrading to Node.js >= 18."
      );
    }
  }
  if (!options.experimental.nodeFetchCompat) {
    options.alias = {
      "node-fetch-native/polyfill": "unenv/runtime/mock/empty",
      "node-fetch-native": "node-fetch-native/native",
      ...options.alias,
    };
  }

  return options;
}

/**
 * @deprecated Please import `defineNitroConfig` from nitropack/config instead
 */
export function defineNitroConfig(config: NitroConfig): NitroConfig {
  return config;
}

export function normalizeRuntimeConfig(config: NitroConfig) {
  provideFallbackValues(config.runtimeConfig);
  const runtimeConfig = defu(config.runtimeConfig, {
    app: {
      baseURL: config.baseURL,
    },
    nitro: {
      envExpansion: config.experimental.envExpansion,
    },
  });
  runtimeConfig.nitro.routeRules = config.routeRules;
  return runtimeConfig as NitroRuntimeConfig;
}

export function normalizeRouteRules(
  config: NitroConfig
): Record<string, NitroRouteRules> {
  const normalizedRules: Record<string, NitroRouteRules> = {};
  for (const path in config.routeRules) {
    const routeConfig = config.routeRules[path] as NitroRouteConfig;
    const routeRules: NitroRouteRules = {
      ...routeConfig,
      redirect: undefined,
      proxy: undefined,
    };
    // Redirect
    if (routeConfig.redirect) {
      routeRules.redirect = {
        to: "/",
        statusCode: 307,
        ...(typeof routeConfig.redirect === "string"
          ? { to: routeConfig.redirect }
          : routeConfig.redirect),
      };
      if (path.endsWith("/**")) {
        // Internal flag
        (routeRules.redirect as any)._redirectStripBase = path.slice(0, -3);
      }
    }
    // Proxy
    if (routeConfig.proxy) {
      routeRules.proxy =
        typeof routeConfig.proxy === "string"
          ? { to: routeConfig.proxy }
          : routeConfig.proxy;
      if (path.endsWith("/**")) {
        // Internal flag
        (routeRules.proxy as any)._proxyStripBase = path.slice(0, -3);
      }
    }
    // CORS
    if (routeConfig.cors) {
      routeRules.headers = {
        "access-control-allow-origin": "*",
        "access-control-allow-methods": "*",
        "access-control-allow-headers": "*",
        "access-control-max-age": "0",
        ...routeRules.headers,
      };
    }
    // Cache: swr
    if (routeConfig.swr) {
      routeRules.cache = routeRules.cache || {};
      routeRules.cache.swr = true;
      if (typeof routeConfig.swr === "number") {
        routeRules.cache.maxAge = routeConfig.swr;
      }
    }
    // Cache: false
    if (routeConfig.cache === false) {
      routeRules.cache = false;
    }
    normalizedRules[path] = routeRules;
  }
  return normalizedRules;
}

function _resolveExportConditions(
  conditions: string[] = [],
  opts: { dev: boolean; node: boolean }
) {
  const resolvedConditions: string[] = [];

  // 1. Add dev or production
  resolvedConditions.push(opts.dev ? "development" : "production");

  // 2. Add user specified conditions
  resolvedConditions.push(...conditions);

  // 3. Add runtime conditions (node or web)
  if (opts.node) {
    resolvedConditions.push("node");
  } else {
    // https://runtime-keys.proposal.wintercg.org/
    resolvedConditions.push(
      "wintercg",
      "worker",
      "web",
      "browser",
      "workerd",
      "edge-light",
      "lagon",
      "netlify",
      "edge-routine",
      "deno"
    );
  }

  // 4. Add default conditions
  resolvedConditions.push("import", "default");

  // Dedup with preserving order
  return resolvedConditions.filter(
    (c, i) => resolvedConditions.indexOf(c) === i
  );
}
